### DISCUSSION TITLE
how AudioStream Apis ensure call sequence safety

### DISCUSSION THREAD

**Discussion by xzhan96** [google/oboe#1793] [Q&A]
As the annotation in AudioStream.h 
![image](https://user-images.githubusercontent.com/13726257/234281261-84e8f05c-3a0b-429d-bf09-c8a00c9ef7b1.png)

All the apis are runs asynchronously, so if I call requestStop before requestStart return success, may the issue happens? like oboe hangs, etc.

**Comment by philburk** [comment]
By "asynchronously", we just mean that the audio will not start flowing immediately. There may be a significant delay, 10-200 msec, before the first callback.

Oboe is meant to be thread safe. But we recommend using a single thread to serialize the open/start/stop/close operations.
This can be complicated because the error callback may stop, close and restart a new stream in its own thread.
You can change this behavior by overriding the onError method:

https://google.github.io/oboe/classoboe_1_1_audio_stream_error_callback.html#a5ad4b8936746ecbb2160a9389b117fc3

**Reply by xzhan96** [reply]
Thanks @philburk ! We have change the behavior by overriding the onError callback. 
However, we met another case sometimes: stuck happens in calling stream_->requestStart(), nothing returns and no voice, no data callback, the only thing we can do is restart the app, in general it works well then. Do you know what are possible reasons of the hang?

**Reply by xzhan96** [reply]
Additional, It seems stuck happens on calling builder.openStream, the builder is oboe::AudioStreamBuilder.
